### PR TITLE
Ticket2017 configurations memory leak

### PR DESCRIPTION
**Pull request by Tom-Willemsen**
### Description of work

This PR fixes a memory leak and associated slowdown when configurations are edited a lot.

The primary changes are:
- Change `Observable` to `ClosableObservable` in various places so that it can be closed.
- Close `ClosableObservables` just before they get garbage collected, using the `finalize()` method.

### To test

https://github.com/ISISComputingGroup/IBEX/issues/2017

Note that there is an occasional exception of the form:

```
Feb 21, 2017 2:20:54 PM org.epics.pvmanager.PVDirector isActive
WARNING: PVReader wasn't properly closed and it was garbage collected. Closing the associated connections...
[stack trace follows]
```

This comes from EPICS, and cannot be caught because it's an exception thrown during garbage collection. A ticket should be created for this exception once this ticket is merged. 

Although I don't like leaving (potential) exceptions in the code, the justification in this case is that this ticket is something that the *users* will benefit from, whereas silencing the exception is something that *as developers* we'd benefit from. Unfortunately fixing that exception is likely to be non-trivial to say the least, and beyond the scope of this ticket I think. The previous solution to this exception was to never let anything be garbage collected in the first place, which is what caused the memory leak!

### Acceptance criteria

There are brief notes on using the java visual VM:
- [ ] [Debugging memory leaks in the IBEX GUI](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Debugging-memory-leaks-in-the-IBEX-GUI)

There is no resource leakage:
- [x] Connect the visual VM. Create 25+ blank configurations, delete them all. Check for a resource leak.
- [ ] [optional] Run through all the manual system tests to do with the configurations panel. Check for a resource leak using the java visual VM. Record the results of these tests if they are performed (see below).

Functionality is unaltered:
- [x] [optional] The manual system tests for configurations pass.
- [x] There are no other functionality changes.

### Example java visual VM screenshots

Healthy example:
image

Unhealthy example:
image

---

#### Code Review

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/GUI-Coding-Conventions)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit tests in place? Are the unit tests small and test the a class in isolation?
- [ ] Are there [system tests](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/System-Testing-with-RCPTT) in place? Do they test a minimal set of functionality and leave the gui as close as possible to its original state?
- [ ] Did any existing system test break as a result of the current changes? 
- [ ] Have the changes been documented in the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev). If so, do they describe the changes appropriately?
- [ ] Has the [manual system tests spreadsheet](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/Manual-system-tests) been updated?
- [ ] If an OPI has been modified, does it conform to the [style guidelines](https://github.com/ISISComputingGroup/ibex_developers_manual/wiki/OPI-Creation)? There is a script called `check_opi_format.py` to help with this.

### Functional Tests

- [ ] Do changes function as described? Add comments below that describe the tests performed.
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has developer documentation been updated if required?

### Final Steps
- [ ] Reviewer has moved the [release notes](https://github.com/ISISComputingGroup/IBEX/wiki/ReleaseNotes_Dev) entry for this ticket in the "Changes merged into master" section

### CODE REVIEW

**Comment by DominicOram** [comment]
I've added the beginnings of a system test (which I used to test the issue). Can you clean it up into a proper test? (https://github.com/ISISComputingGroup/ibex_system_tests/tree/Ticket2017_configurations_memory_leak/RCPTT_Tests)

File: base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/writing/ForwardingWritable.java [hunk 1]
```diff
@@ -24,6 +24,14 @@
 import uk.ac.stfc.isis.ibex.epics.observing.Subscription;
 import uk.ac.stfc.isis.ibex.epics.pv.Closable;
 
+/**
+ * A forwarding writable.
+ *
+ * @param <TIn>
```
Comment (DominicOram) [inline base/uk.ac.stfc.isis.ibex.epics/src/uk/ac/stfc/isis/ibex/epics/writing/ForwardingWritable.java]: Can you be a bit more descriptive with Tin/Tout descriptions?

**Comment by DominicOram** [comment]
Small comment in the code otherwise looks good. Can you add exploring the exception as a new ticket please?

**Comment by Tom-Willemsen** [comment]
Exploring the exception has been added as ticket 2131.

**Comment by DominicOram** [comment]
Can you resolve conflicts? Then I will merge

**Pull request merged**